### ISSUE TITLE
ARM / Win10 IoT -- crashes when code contracts added

### ISSUE THREAD

**Issue by henrygab** [Microsoft/CodeContracts#253]
### Questions:
1. Should one expected to be able to include Code Contracts in a Windows 10 UWP application?
2. Presuming yes, should one still expect them to work, where the target is Raspberry Pi 2 (running Win10 IoT)?
3. Presuming they should work on Win10 IoT, is the issue below a known bug?
4. Presuming it’s not a known bug, is there anything I can do to help the team track down and root-cause this issue, so it can be triaged and potentially fixed?
 
Note: I've created a minimal example project, so I can create a GitHub depot if requested.

### Issue:
Adding Code Contracts to a Windows 10 UWP App (targeting Windows 10 IoT on the Raspberry Pi 2, ARM) causes application corruption / crashes [100% repro]

### Steps to repro:
This may have more steps than needed.  Example source available.
* Using Visual Studio 2015...
* Create a simple headless, task-based app/service for IoT
	* Include at least one additional class (the target class) that is instantiated by the task
	* Have one public constructor with a first parameter being a class type, a second parameter being a struct type
	* See sample code block 1:
* Load the UWP App on the RPi2 (under debugger)
* Verify the app loads and runs as expected on the RPi2, including creating the class, etc.
* Toss into source control for later comparison
* Add some simple code contracts.
    * See sample code block 2
* Load the UWP App on the RPi2 (under debugger)

```
    // Code Block 1
    using K = System.Diagnostics.CodeContracts;
    using IIC = Windows.Devices.I2C;
        public class TestClass {
            private class _Unusable { private _Unusable(); }
            private TestClass(IIC.Device device, byte v, _Unusable mustBeNull) {
                if (null == device) { throw new ArgumentNullException("device");
                m_Device = device;
            }
            public TestClass(IIC.Device device)
                : this(device, 0x04, (_Unusable)null) {}
    }
```

```
            // Code Block 2
            K.Contracts.Requires(null == mustBeNull);
            K.Contracts.Requires(null != device);
            K.Contracts.Requires((0 <= v) && (7 >= v)); // yes, I know first check is always true…
```

### Expected Results:
The application continues to run perfectly.

### Actual Results:
App starts, and runs fine until calling the class constructor that had Code Contracts added to it.
At that point, things go haywire.  No catchable exception occurs.  Often, the program simply terminates.  Twice, stepping into the constructor caused the app to appear to restart (instruction pointer corruption??).

**Comment by SergeyTeplyakov** [comment]
Is it possible to look at compiled IL? Or Windows 10 UAP apps are using native compilation?

**Comment by henrygab** [comment]
I'm working to provide some additional information.

**Comment by henrygab** [comment]
I cannot commit to quickly providing the time to investigate.  Please feel free to close if having this open for a month or two is not quick enough....

**Comment by henrygab** [comment]
I now believe the issue related to my inexperience with Win10 IOT.  :)  Thank you for allowing the issue to stay open, even when I could not commit to a quick follow-up.
Up soon: 2-3 new issues related to interface contracts silently being ignored... this time with source from the get-go.

**Comment by SergeyTeplyakov** [comment]
@henrygab I would glad to hear, what led to those issues... Just out of curiosity... :)

**Comment by henrygab** [comment]
I had a bug where the task was being prematurely terminated... Or at least I think that was the bug.  I rolled back to earlier version of the code, verified it was OK, and re-added the contacts without issue.

So, it wasn't hardware, software, our firmware, but a wetware bug (aka pebkas, aka ID:10T error 😊).

Specifically, I think I was trying to"play it safe", by ensuring I terminated my task, but failed to wait on the task completion.  Given the use of async, which is still fairly new to me, the termination of the task was attempted before I really wanted it to be.  <sheepish grin>

But don't worry. I've added two more bugs that I think may be •real• issues for interface contract inheritance using ContractClassFor....  😊

Sent from my Windows Phone
